src/ms3/bs4_parser.py

```diff
@@ -2004,13 +2004,99 @@ def __getstate__(self):
 #######################################################################
 
+class Parts(LoggedClass):
+    def __init__(self, soup: bs4.BeautifulSoup, **logger_cfg):
+        super().__init__('Parts', logger_cfg)
+        self.parts_data = {f"part_{i}": part for i, part in enumerate(soup.find_all('Part'), 1)}
+
+    @property
+    def map_staff2part(self) -> dict[list, str]:
+        """Returns the dict in the format {[2, 3]: 'part_1'} for staves 2 and 3 of part 1"""
+        staff2part = {}
+        for key_part, part in self.parts_data.items():
+            staves = [f"staff_{staff['id']}" for staff in part.find_all('Staff')]
+            staff2part.update(dict.fromkeys(staves, key_part))
+        return staff2part
+
+
 class Instrumentation(LoggedClass):
     """Easy way to read and write the instrumentation of a score, that is
     'instrument', 'longName', 'shortName', 'trackName', 'instrumentId'."""
 
     def __init__(self, soup: bs4.BeautifulSoup, **logger_cfg):
         super().__init__('Instrumentation', logger_cfg)
+        self.INSTRUMENT_DEFAULTS = {
+            'Harpsichord': {'instrumentId': 'keyboard.harpsichord',
+                            'longName': 'Harpsichord',
+                            'part_trackName': 'Harpsichord',
+                            'shortName': 'Hch.',
+                            'trackName': 'Harpsichord'},
+            'Piano': {'instrumentId': 'keyboard.piano',
+                      'longName': 'Piano',
+                      'part_trackName': 'Piano',
+                      'shortName': 'Pno.',
+                      'trackName': 'Piano'},
+            'Violoncello': {'instrumentId': 'strings.cello',
+                            'longName': 'Violoncello',
+                            'part_trackName': 'Cello',
+                            'shortName': 'Vc.',
+                            'trackName': 'Violoncello'},
+        }
+        self.part_tracknames = [elem['part_trackName'] for elem in self.INSTRUMENT_DEFAULTS.values()]
         self.soup = soup
+        # 'instrument',
+        self.fields_names = ['longName', 'shortName', 'trackName', 'instrumentId', 'part_trackName']
+        self.parts = Parts(soup)
+        self.text_tags = self.text_tags_fn()  # store references to XML tags
+
+
+    def text_tags_fn(self) -> dict[str, dict[str, bs4.Tag]]:
+        """Returns the dict of self.fields_names info for every part  {[staff_2, staff_3]: 'part_1'} for staves 2 and 3 of part 1"""
+        tag_dict = {}
+        for key_part, part in self.parts.parts_data.items():
+            instrument_info = part.Instrument
+            staves = [f"staff_{(staff['id'])}" for staff in part.find_all('Staff')]
+            cur_dict = {"instrumentId": instrument_info["id"]}
+            for name in self.fields_names:
+                if name == "part_trackName":
+                    part_tracknames = part.find_all("trackName")
+                    tags = list(filter(lambda k: k.get_text() in self.part_tracknames,
```

Comment:
Here you assume that a <Part> can include several <trackName> tags and filter out those that are (already) included in our default instruments. I don't think that this is the correct logic.

AFAIK, each <Part> has only a single <trackName> or none (please correct me if I'm wrong). The logic goes that you use the one that is present (`part.find('trackName')`, which yields `None` if not present), check if it's included in `part_tracknames`, and throw a `NotImplementedError` otherwise (or some other way to deal with it during the development process).

Reply:
You are right, there is always 1 part_trackName among the files presented in the test. With the new tests it is not necessary to check for the presence in the fixed spoke. So in the new version I take the value directly.